### PR TITLE
Feature/ccp 253

### PR DESCRIPTION
**Pull request by kkrish39**
- I pulled the SQL migration branch changes in-order to be on the same page. Hence, Integration tests will still be failing.

- for home.html, I applied indentation because of which it shows that the entire file is changed.

### CODE REVIEW

File: Conceptpower+Spring/src/main/java/edu/asu/conceptpower/app/core/IRequestsManager.java [hunk 1]
```diff
@@ -47,4 +47,9 @@
      * @return List<ReviewRequest> - return all the reviewRequests in the database.
      */
     public List<ReviewRequest> getAllReviews();
+    
+    /**
+     * @return List<ReviewRequest> - return all the reviewRequests that are oven.
```
Comment (jdamerow) [inline Conceptpower+Spring/src/main/java/edu/asu/conceptpower/app/core/IRequestsManager.java]: "open" (they are not in the oven 😄)

File: Conceptpower+Spring/src/main/java/edu/asu/conceptpower/app/model/ChangeEvent.java [hunk 2]
```diff
@@ -45,9 +43,8 @@ public ChangeEvent(String userName, Date date, ChangeEventTypes type) {
     @GeneratedValue(strategy=GenerationType.IDENTITY)
     private int id;
     
-    @ManyToOne
-    @JoinColumn(name = "concept_entry_id", referencedColumnName="concept_id")
-    private ConceptEntry concept;
+    @Column(name = "concept_entry_id")
+    private String concept_entry_id;
```
Comment (jdamerow) [inline Conceptpower+Spring/src/main/java/edu/asu/conceptpower/app/model/ChangeEvent.java]: Should follow our code style.

File: Conceptpower+Spring/src/main/java/edu/asu/conceptpower/app/model/ChangeEvent.java [hunk 3]
```diff
@@ -91,12 +88,12 @@ public void setType(ChangeEventTypes type) {
         this.type = type;
     }
     
-    public ConceptEntry getConcept() {
-        return concept;
+    public String getConcept() {
```
Comment (jdamerow) [inline Conceptpower+Spring/src/main/java/edu/asu/conceptpower/app/model/ChangeEvent.java]: getter should be named correctly

File: Conceptpower+Spring/src/main/java/edu/asu/conceptpower/app/model/Comment.java [hunk 4]
```diff
@@ -35,15 +36,15 @@
     @Column(name="created_at")
     private OffsetDateTime createdAt;
     
-    @ManyToOne
-    @JoinColumn(name = "review_request_id", referencedColumnName="concept_id")
-    private ReviewRequest review;
+    @Column(name="review_id")
+    private String review_id;
```
Comment (jdamerow) [inline Conceptpower+Spring/src/main/java/edu/asu/conceptpower/app/model/Comment.java]: needs to follow our code style

File: Conceptpower+Spring/src/main/java/edu/asu/conceptpower/web/ReviewRequestController.java [hunk 5]
```diff
@@ -33,18 +34,24 @@
...
         return requestsMgr.updateReview(reviewId, ReviewStatus.OPENED, comment, OffsetDateTime.now(), principal.getName());
     }
     
-    @RequestMapping(value = "/auth/request/{conceptId}/all", method = RequestMethod.GET)
+    @GetMapping(value = "/auth/request/{conceptId}/all")
     public @ResponseBody List<ReviewRequest> getAllReviews(@PathVariable String conceptId, Principal principal){
         return requestsMgr.getAllReviews(conceptId);  
     }
+    
+    @GetMapping(value="/auth/openrequests")
```
Comment (jdamerow) [inline Conceptpower+Spring/src/main/java/edu/asu/conceptpower/web/ReviewRequestController.java]: should follow the other url patterns, e.g. `/auth/request/open` or `/auth/request/all/open`

File: Conceptpower+Spring/src/main/java/edu/asu/conceptpower/web/ReviewRequestController.java [hunk 6]
```diff
@@ -33,18 +34,24 @@
...
+    public String viewOpenReviewRequests(ModelMap model) {
+        model.addAttribute("openRequests", requestsMgr.getAllOpenReviews());
+        return "layouts/concepts/openrequests";
+    }
```
Comment (jdamerow) [inline Conceptpower+Spring/src/main/java/edu/asu/conceptpower/web/ReviewRequestController.java:65]: should be in it's own controller (usually, you have one controller per page)

File: Conceptpower+Spring/src/main/webapp/WEB-INF/views/layouts/concepts/openrequests.html [hunk 7]
```diff
@@ -0,0 +1,168 @@
...
+   >
+   <head>
+      <link th:href="@{/resources/assets/css/reviewrequest.css}" rel="stylesheet" />
+      <script>
+      	$(document).on('click', '#conceptDetails', function(e) {
+    	    var conceptid = $(this).data('conceptid');
+
+    	    $.ajax({
+    	        type: "GET",
+    	        url: "[[@{'/conceptDetail'}]]",
```
Comment (jdamerow) [inline Conceptpower+Spring/src/main/webapp/WEB-INF/views/layouts/concepts/openrequests.html]: I believe you dont' need the `"`; Thymeleaf should take care of that when you use `[[...]]` (instead of `[(...)]`

File: Conceptpower+Spring/src/main/webapp/WEB-INF/views/layouts/concepts/openrequests.html [hunk 8]
```diff
@@ -0,0 +1,168 @@
...
+    	            $("#conceptDetailsType").text(details.type);
+    	            $("#conceptDetailsEqualTo").text(details.equalto);
+    	            $("#conceptDetailsSimilarTo").text(details.similarto);
+    	            $("#conceptDetailsCreator").text(details.creator);
+    	            $("#conceptDetailsDesc").html(details.description);
+    	            $("#conceptDetailsMergedIds").html(details.mergedIds);
+
+    	            if (details.mergedIds) {
+    	                $.ajax({
+    	                    url: "[[@{'/rest/OriginalConcepts'}]]",
```
Comment (jdamerow) [inline Conceptpower+Spring/src/main/webapp/WEB-INF/views/layouts/concepts/openrequests.html]: same here

**Review by jdamerow — CHANGES_REQUESTED** [review]

File: Conceptpower+Spring/src/main/webapp/WEB-INF/views/layouts/concepts/openrequests.html (hunk 7 above)
Comment (kkrish39) [reply Conceptpower+Spring/src/main/webapp/WEB-INF/views/layouts/concepts/openrequests.html]: I think it's true if we use it inside the HTML. Using it in jQuery without the double quotes raises an error.

Comment (jdamerow) [reply Conceptpower+Spring/src/main/webapp/WEB-INF/views/layouts/concepts/openrequests.html]: Are you sure? Have you tried it? This says otherwise:
https://www.thymeleaf.org/doc/tutorials/3.0/usingthymeleaf.html (12.3 JavaScript inlining)

Would be good to know if that is wrong.

Comment (jdamerow) [reply Conceptpower+Spring/src/main/webapp/WEB-INF/views/layouts/concepts/openrequests.html]: Actually, never mind. Apparently for links things work a little different (https://github.com/thymeleaf/thymeleaf/issues/84)

Comment (kkrish39) [reply Conceptpower+Spring/src/main/webapp/WEB-INF/views/layouts/concepts/openrequests.html]: Because when I try without the double quote, I'm clearly getting an error in the console.

Uncaught SyntaxError: Invalid regular expression flags

File: Conceptpower+Spring/src/main/java/edu/asu/conceptpower/app/manager/RequestsManager.java [hunk 9]
```diff
@@ -113,4 +113,9 @@ protected String generateId(String prefix) {
     public List<ReviewRequest> getAllReviews() {
        return dbClient.getAllReviews();
     }
+
+    @Override
+    public List<ReviewRequest> getAllOpenReviews() {
```
Comment (jdamerow) [inline Conceptpower+Spring/src/main/java/edu/asu/conceptpower/app/manager/RequestsManager.java]: Just make that return reviews by status and then call it whereever you need all reviews with a specific status.

File: Conceptpower+Spring/src/main/java/edu/asu/conceptpower/web/ViewOpenReviewRequestsController.java [hunk 10]
```diff
@@ -0,0 +1,27 @@
...
+public class ViewOpenReviewRequestsController {
+	@Autowired
+    private RequestsManager requestsMgr;
+
+	@GetMapping(value="/auth/request/all/open")
+    public String viewOpenReviewRequests(ModelMap model) {
+        model.addAttribute("openRequests", requestsMgr.getAllOpenReviews());
+        return "layouts/concepts/openrequests";
+    }
+}
```
Comment (jdamerow) [inline Conceptpower+Spring/src/main/java/edu/asu/conceptpower/web/ViewOpenReviewRequestsController.java:85]: formatting

File: Conceptpower+Spring/src/main/java/edu/asu/conceptpower/web/ViewOpenReviewRequestsController.java [hunk 11]
```diff
@@ -0,0 +1,27 @@
...
+
+@Controller
+public class ViewOpenReviewRequestsController {
+	@Autowired
+    private RequestsManager requestsMgr;
+
+	@GetMapping(value="/auth/request/all/open")
+    public String viewOpenReviewRequests(ModelMap model) {
+        model.addAttribute("openRequests", requestsMgr.getAllOpenReviews());
+        return "layouts/concepts/openrequests";
```
Comment (jdamerow) [inline Conceptpower+Spring/src/main/java/edu/asu/conceptpower/web/ViewOpenReviewRequestsController.java]: I'm confused, the story is about "list of all concepts with open review requests", but you're returning requests here. The list of concepts should pretty much look like the list of search results, which from looking at the template is not the case?

**Review by jdamerow — CHANGES_REQUESTED** [review]

File: Conceptpower+Spring/src/main/java/edu/asu/conceptpower/web/ViewOpenReviewRequestsController.java (hunk 11 above)
Comment (kkrish39) [reply Conceptpower+Spring/src/main/java/edu/asu/conceptpower/web/ViewOpenReviewRequestsController.java]: Ahh yes, Julia. I'm actually querying ReviewRequest Table with ```ReviewStatus.OPENED```. The new page looks like this. We can view the concept details by clicking on the concept I'd (A modal will popup)

Screen Shot 2020-11-05 at 10 32 42 AM

Comment (kkrish39) [reply Conceptpower+Spring/src/main/java/edu/asu/conceptpower/web/ViewOpenReviewRequestsController.java]: Let me know if this is ok. If not, I'll join and return the concept entry table and render the complete data.

Comment (jdamerow) [reply Conceptpower+Spring/src/main/java/edu/asu/conceptpower/web/ViewOpenReviewRequestsController.java]: We need more info than that (pretty much what's in the search results). This doesn't tell the user anything about the concepts, which is the key point of interest here.

**Comment by kkrish39** [comment]
The Updated UI
Screen Shot 2020-11-05 at 3 44 17 PM

**Comment by jdamerow** [comment]
Resolve conflicts please

File: Conceptpower+Spring/src/main/java/edu/asu/conceptpower/app/manager/impl/ConceptManager.java [hunk 12]
```diff
@@ -628,4 +634,15 @@ public ConceptEntry getOriginalConceptEntry(String id) {
...
         return client.getAllConcepts();
     }
+
+	@Override
+	public List<ConceptEntry> getAllConceptsByStatus(ReviewStatus status) {
+		List<ConceptEntry> result = new ArrayList<>();
+		
+		for(ReviewRequest request: requestsManager.getReviewByStatus(status)) {
+			result.add(getConceptEntry(request.getConceptId()));
+		}
```
Comment (jdamerow) [inline Conceptpower+Spring/src/main/java/edu/asu/conceptpower/app/manager/impl/ConceptManager.java]: have you tried using a join query, something like
`SELECT c FROM ConceptEntry c, ReviewRequest r WHERE c.id = r.concept_id AND r.status = :status`?

I'm really not an expert on this, but I"m sure there is a way to let the database do all of this and avoid multiple database calls, especially since you'll want to paginate.

File: Conceptpower+Spring/src/main/java/edu/asu/conceptpower/web/ViewOpenReviewRequestsController.java [hunk 13]
```diff
@@ -0,0 +1,67 @@
...
+		
+		List<ConceptEntry> openReviewConcepts = conceptsManager.getAllConceptsByStatus(ReviewStatus.OPENED);
+		List<ConceptEntryWrapper> foundConcepts = null;
+		
+		if (indexService.isIndexerRunning()) {
+            model.addAttribute("show_error_alert", true);
+            model.addAttribute("error_alert_msg", indexerRunning);
+            
+            return "layouts/concepts/openrequests";
+        }
```
Comment (jdamerow) [inline Conceptpower+Spring/src/main/java/edu/asu/conceptpower/web/ViewOpenReviewRequestsController.java]: this should be first in your method. No need to query the db if the results are not displayed.

File: Conceptpower+Spring/src/main/java/edu/asu/conceptpower/web/ViewOpenReviewRequestsController.java [hunk 14]
```diff
@@ -0,0 +1,67 @@
...
+		
+		
+        try {
+			foundConcepts = wrapperCreator.createWrappers(openReviewConcepts.toArray(new ConceptEntry[openReviewConcepts.size()]));
+		} catch (LuceneException e) {
+			model.addAttribute(indexerStatus, e.getMessage());
+			e.printStackTrace();
```
Comment (jdamerow) [inline Conceptpower+Spring/src/main/java/edu/asu/conceptpower/web/ViewOpenReviewRequestsController.java:76]: tststs. needs handling ;)

File: Conceptpower+Spring/src/main/java/edu/asu/conceptpower/web/ViewOpenReviewRequestsController.java (hunk 10 above)
Comment (jdamerow) [reply Conceptpower+Spring/src/main/java/edu/asu/conceptpower/web/ViewOpenReviewRequestsController.java:85]: !

File: Conceptpower+Spring/src/main/webapp/WEB-INF/views/layouts/concepts/openrequests.html [hunk 15]
```diff
@@ -0,0 +1,183 @@
...
+    	            success: function(response) {
+    	                window.location = "[[@{'/auth/request/all/open'}]]"
+    	            },
+    	            error: function(e) {
+    	                $('#resolveCommentError').text("Something went wrong. Unable to resolve the request");
+    	                $('#resolveCommentError').show();
+    	            }
+    	        });
+    	    }
+    	});
```
Comment (jdamerow) [inline Conceptpower+Spring/src/main/webapp/WEB-INF/views/layouts/concepts/openrequests.html:43]: a lot of the javascript is the same for home and this page, right? so it should be in an external file.

Comment (jdamerow) [reply Conceptpower+Spring/src/main/webapp/WEB-INF/views/layouts/concepts/openrequests.html:43]: DRY

**Review by jdamerow — CHANGES_REQUESTED** [review]

File: Conceptpower+Spring/src/main/java/edu/asu/conceptpower/app/manager/impl/ConceptManager.java (hunk 12 above)
Comment (kkrish39) [reply Conceptpower+Spring/src/main/java/edu/asu/conceptpower/app/manager/impl/ConceptManager.java]: You are right. But the trick is that not all the concepts are in our MySQL database. We have concepts coming from both the MySQL and the wordnet DB.

Comment (jdamerow) [reply Conceptpower+Spring/src/main/java/edu/asu/conceptpower/app/manager/impl/ConceptManager.java]: I think we need to change the way reviews can be created then. Basically, before a someone can add a review request to a concept that concept has to be turned into a "real" concept (conceptwrapper).

File: Conceptpower+Spring/src/main/java/edu/asu/conceptpower/web/ViewOpenReviewRequestsController.java (hunk 14 above)
Comment (jdamerow) [reply Conceptpower+Spring/src/main/java/edu/asu/conceptpower/web/ViewOpenReviewRequestsController.java:76]: ?

File: Conceptpower+Spring/src/main/java/edu/asu/conceptpower/web/ViewOpenReviewRequestsController.java [hunk 16]
```diff
@@ -0,0 +1,69 @@
...
+	private IIndexService indexService;
+
+	@Value("#{messages['INDEXER_RUNNING']}")
+	private String indexerRunning;
+
+	@Value("#{messages['INDEXERSTATUS']}")
+	private String indexerStatus;
+
+	@GetMapping(value = "/auth/request/all/open")
+	public String viewOpenReviewRequests(ModelMap model) {
```
Comment (jdamerow) [inline Conceptpower+Spring/src/main/java/edu/asu/conceptpower/web/ViewOpenReviewRequestsController.java]: this needs to be paginated

**Review by jdamerow — CHANGES_REQUESTED** [review]

**Comment by kkrish39** [comment]
Few changes weren't reflected. To check, re-opened the PR and marked it as draft.

File: Conceptpower+Spring/src/main/java/edu/asu/conceptpower/app/manager/impl/ConceptManager.java [hunk 17]
```diff
@@ -640,4 +646,31 @@ public ConceptEntry getOriginalConceptEntry(String id) {
...
+		
+		return result;
+    }
+    
+	@Override
+	public Integer getNumberOfConceptsByStatus(ReviewStatus status) {
+		return requestsManager.getReviewByStatus(status).size();
+	}
```
Comment (jdamerow) [inline Conceptpower+Spring/src/main/java/edu/asu/conceptpower/app/manager/impl/ConceptManager.java]: indentation

File: Conceptpower+Spring/src/main/java/edu/asu/conceptpower/app/manager/impl/ConceptManager.java (hunk 12 above)
Comment (jdamerow) [reply Conceptpower+Spring/src/main/java/edu/asu/conceptpower/app/manager/impl/ConceptManager.java]: ?

**Review by jdamerow — CHANGES_REQUESTED** [review]
previous comments don't seem to be addressed?

**Comment by kkrish39** [comment]
Oh yes, Julia. I just reopened as few previous changes were not reflected. I commented on top. I will push the changes in a few minutes.

**Comment by kkrish39** [comment]
One more change needs to be done. Closing the PR.

File: Conceptpower+Spring/src/main/java/edu/asu/conceptpower/app/db/DatabaseClient.java [hunk 18]
```diff
@@ -305,4 +305,19 @@ private String getLikePattern(final String searchTerm) {
           
         };
       }
+
+	@Override
+	public List<ConceptEntry> getAllConceptsByStatus(ReviewStatus status) {
```
Comment (jdamerow) [inline Conceptpower+Spring/src/main/java/edu/asu/conceptpower/app/db/DatabaseClient.java]: is this still used?

File: Conceptpower+Spring/src/main/java/edu/asu/conceptpower/app/db/DatabaseClient.java [hunk 19]
```diff
@@ -305,4 +305,19 @@ private String getLikePattern(final String searchTerm) {
...
+	}
+	
+	@Override
+	public List<ConceptEntry> getAllConceptsByStatusPaginated(ReviewStatus status, Integer page, Integer pageSize) {
+		return conceptEntryRepository.getConceptsForGivenStatus(status, PageRequest.of(page, pageSize));
+	}
+	
+	@Override
+	public Integer getNumberOfConceptsByStatus(ReviewStatus status) {
+		return conceptEntryRepository.getConceptsForGivenStatus(status).size();
```
Comment (jdamerow) [inline Conceptpower+Spring/src/main/java/edu/asu/conceptpower/app/db/DatabaseClient.java]: this should use a count method to have the db do the counting and not load all objects into memory

File: Conceptpower+Spring/src/main/java/edu/asu/conceptpower/app/manager/IConceptManager.java [hunk 20]
```diff
@@ -202,4 +203,27 @@ public int getPageCountForConceptEntries(String word, String pos, String concept
...
+    
+    /**
+     * 
+     * @param status - required review status entries
+     * @return - List<ConceptEntry> that matches the given status
+     */
+    public List<ConceptEntry> getAllConceptsByStatus(ReviewStatus status);
+    
+    /**
+     *
```
Comment (jdamerow) [inline Conceptpower+Spring/src/main/java/edu/asu/conceptpower/app/manager/IConceptManager.java:209]: needs a description

File: Conceptpower+Spring/src/main/java/edu/asu/conceptpower/web/ViewOpenReviewRequestsController.java [hunk 21]
```diff
@@ -0,0 +1,82 @@
...
+		if (indexService.isIndexerRunning()) {
+			model.addAttribute("show_error_alert", true);
+			model.addAttribute("error_alert_msg", indexerRunning);
+
+			return "layouts/concepts/openrequests";
+		}
+		
+		int pageNo = page == null ? 0 : Integer.parseInt(page);
+		int numRecordsPerPage = pageSize == null ? defaultNumRecordsPerPage : Integer.parseInt(pageSize);
+		int totalPageCount = (int) Math.ceil((double)conceptsManager.getNumberOfConceptsByStatus(ReviewStatus.OPENED)/(double) defaultNumRecordsPerPage);
```
Comment (jdamerow) [inline Conceptpower+Spring/src/main/java/edu/asu/conceptpower/web/ViewOpenReviewRequestsController.java]: should that be dividing by `numRecordsPerPage`?

File: Conceptpower+Spring/src/main/java/edu/asu/conceptpower/web/ViewOpenReviewRequestsController.java [hunk 22]
```diff
@@ -0,0 +1,82 @@
...
+
+			return "layouts/concepts/openrequests";
+		}
+		
+		int pageNo = page == null ? 0 : Integer.parseInt(page);
+		int numRecordsPerPage = pageSize == null ? defaultNumRecordsPerPage : Integer.parseInt(pageSize);
+		int totalPageCount = (int) Math.ceil((double)conceptsManager.getNumberOfConceptsByStatus(ReviewStatus.OPENED)/(double) defaultNumRecordsPerPage); 
+		
+		pageNo = pageNo < 0 ? 0 : pageNo;
+		numRecordsPerPage = numRecordsPerPage < 0 ? defaultNumRecordsPerPage : numRecordsPerPage;
```
Comment (jdamerow) [inline Conceptpower+Spring/src/main/java/edu/asu/conceptpower/web/ViewOpenReviewRequestsController.java]: why is that here again?

File: Conceptpower+Spring/src/main/java/edu/asu/conceptpower/web/ViewOpenReviewRequestsController.java (hunk 14 above)
Comment (jdamerow) [reply Conceptpower+Spring/src/main/java/edu/asu/conceptpower/web/ViewOpenReviewRequestsController.java:76]: ?

File: Conceptpower+Spring/src/main/java/edu/asu/conceptpower/web/ViewOpenReviewRequestsController.java (hunk 10 above)
Comment (jdamerow) [reply Conceptpower+Spring/src/main/java/edu/asu/conceptpower/web/ViewOpenReviewRequestsController.java:85]: ?

**Review by jdamerow — CHANGES_REQUESTED** [review]
Please reread the pull request guidelines in confluence.

File: Conceptpower+Spring/src/main/java/edu/asu/conceptpower/web/ViewOpenReviewRequestsController.java (hunk 22 above)
Comment (kkrish39) [reply Conceptpower+Spring/src/main/java/edu/asu/conceptpower/web/ViewOpenReviewRequestsController.java]: We were checking only for null condition before

File: Conceptpower+Spring/src/main/java/edu/asu/conceptpower/web/ViewOpenReviewRequestsController.java (hunk 14 above)
Comment (kkrish39) [reply Conceptpower+Spring/src/main/java/edu/asu/conceptpower/web/ViewOpenReviewRequestsController.java:76]: Ahh..I pushed the changes. I'm not sure why it's not getting reflected here. Let me check. This is the code in my local.
```
              try {
		    if(!openReviewConcepts.isEmpty()) {
		        foundConcepts = wrapperCreator
		                .createWrappers(openReviewConcepts.toArray(new ConceptEntry[openReviewConcepts.size()]));
		    }
		} catch (LuceneException e) {
			model.addAttribute(indexerStatus, e.getMessage());
			e.printStackTrace();
		}

		model.addAttribute("openRequests", foundConcepts);
		model.addAttribute("page",pageNo);
		model.addAttribute("count",totalPageCount);
		
		return "layouts/concepts/openrequests";
```

**Comment by jdamerow** [comment]
there seem to be several unadressed comments

**Comment by kkrish39** [comment]
I can see that few changes are not reflected in the PR. But if I open the file separately I can find those. Please let me know even then you find any PR comments I missed addressing.

**Comment by jdamerow** [comment]
please reopen a new pr. I can't hunt down all possible changes that the PR doesn't show (please reread our code review guidelines about valuing your reviewers time ;)
